### PR TITLE
Update Google Sign In

### LINKED ISSUE #17429: Migrate away from deprecated Google Platform Library
Google is deprecating the Google Platform Library (see: https://developers.googleblog.com/2021/08/gsi-jsweb-deprecation.html), and intends to fully retire by March 2023. Metabase should migrate towards the Google Identity Services API (https://developers.googleblog.com/2021/07/launching-our-new-google-identity-services-apis.html) to avoid metabase instances from breaking unexpectedly.

### PR DESCRIPTION
**Pull request by gusaiani**
Fixes #17429

Please see issue above for justification and details.

tl;dr: New Metabase users logging in with Google Sign In may not be able to do so until we upgrade Google's Sign In library.

🎗️ 
There's more to do here, but this PR should do the job if we decide to ship very soon, or be able to test with users very soon.
In my tests it's been working fine.

### Advantages of this solution

Implementation is really straightforward.
We may be able to clean up a lot of code and remove a lot of indirection if we use this.

We may later quickly switch to the "one tap" version of the button, which should make signing in faster: button will suggest an email account to use, should the user have more than one Google emails. Plus, developers can add tips on what that email should be.

### Trade-offs to note

1. Button is now the official Google button, not a customized button.
We can't do drop shadows or vertical padding, for example.
Button look can be customized under Google's parameters.

2. Some unit tests have been skipped for agility, and can be rethought if we go with this solution.

3. Metabase users with languages other than English will see a flash of "Sign in With Google" before they see the same copy in their language.

## How to Test

1. Go to http://localhost:3000/admin/settings/authentication/google
2. Add Google Client ID [you can find here](https://metaboat.slack.com/archives/C03E3Q5T4P4/p1653085210635789)
3. Log out
4. Log in with Google 

You should be able to log in.

### CODE REVIEW

**Review by alxnddr — COMMENTED** [review]
Can you please add test steps? Auth locally did not work for me, maybe I've done something wrong

**Comment by gusaiani** [comment]
> Can you please add test steps? Auth locally did not work for me, maybe I've done something wrong

@alxnddr, oops, I forgot how not-obvious this is.

Added in PR description above.

**Review by flamber — COMMENTED** [review]
When testing with my existing setups, then I'm just seeing a blank popup with a browser log `[GSI_LOGGER]: The given origin is not allowed for the given client ID.`
image

The temporary quickfix for the issue is to add `plugin_name` https://github.com/metabase/metabase/issues/17429#issuecomment-1126884333. Then that can be shipped in the next patch release.

That will give us a little time to completely rework how our Google implementation should be in 44, which should just be the one-tab solution that Google also recommends, since it would avoid a lot of the problems we currently have with the popup.

**Comment by gusaiani** [comment]
@flamber 

The origins I set up for the clientID I shared are

http://localhost:3000
http://localhost/auth/signin (or the URL for sign in, I'm afk now)

If you are using the clientID I shared, first thing I would check if your setup does localhost:3000

Alternatively you can create your own Google app and customize it.

At first sight this looks like a configuration thing.

**Comment by flamber** [comment]
@gusaiani Okay, what are the steps to setup authentication on https://console.developers.google.com/, since I'm guessing it is different than currently, so that has to be very explicitly noted, since it would not be a drop-in-replacement and break peoples auth on upgrade.
Google Sign-in is working fine on 0.43.1 with my current test setups. I'm using my own credentials.

**Comment by mazameli** [comment]
I left the `Domain` input box in blank, and tried logging in with a gmail account (`…@gmail.com`) that matched the email addess of a user that I manually added. My expectation is that I'm supposed to be able to log in with that user via GAuth now, but I just saw a blank auth popover after going through the google sign-in flow and doing 2FA:

image

image


Much more minor, it would be cool if we could get the button styed to look like how it currently does in production:

image

**Comment by mazameli** [comment]
I tried it again after adding `gmail.com` to the `Domains` setting, and the same thing happened: blank auth popover after trying to sign in and confirming via 2FA.

**Comment by gusaiani** [comment]
@mazameli, thanks for the review.

There's a bit around styling this button now with the Google lib update.

The new button Google offers out of the box can be a few different things.
As one example, it can be a "one-tap" sign in or sign up, suggesting an email for the person to just do it with.
See [docs](https://developers.google.com/identity/gsi/web).

However, the button is really an iframe, so it can't be styled as a button could.

Alternatively, we can create a fully customized button and make it look the way we want.
If we go this way, and this is indeed weird, Google will return completely different data on a successful sign up.
This will require us to write code on the backend to process this data and complete the sign in.
And we will lose the ability to switch to "one-tap" in the future, or switch and maintain the styling.

So if this is a minor consideration, I suggest we start by using the Google look.

**Review by cdeweyx — APPROVED** [review]

**Comment by gusaiani** [comment]
@mazameli, I've just tested it again, including 2FA.

It works all right on my machine.

I think, because of the way the Google sign in application works, maybe if you're using another port locally or some other detail, this may not work — you'd need to create your own application with your own domains aligned with what you're using.

tl;dr: technically this seems to be working ok, we need to be sure our users have the Google console apps created correctly (and this includes ourselves on our internal use of Metabase)

**Comment by mazameli** [comment]
@gusaiani okay, cool, thanks for letting me know. If others have tested this and it works, it could very well be something to do with the way I'm running things locally (I'm just running it off the branch with `yarn run build-hot` and `clojure -M:run:ee`, not off of a built jar), and I definitely did not "create my own application with my own domains aligned with what I'm using."

File: src/metabase/server/middleware/security.clj [hunk 1]
```diff
@@ -51,7 +51,7 @@
                                   ["'self'"
                                    "'unsafe-eval'" ; TODO - we keep working towards removing this entirely
                                    "https://maps.google.com"
-                                   "https://apis.google.com"
+                                   "https://accounts.google.com"
```
Comment (flamber) [inline src/metabase/server/middleware/security.clj:54]: Are the other google/gstatic/googleapis even needed anymore?

**Review by flamber — COMMENTED** [review]
@gusaiani So the same steps as currently? https://www.metabase.com/docs/latest/administration-guide/10-single-sign-on.html#enabling-google-sign-in

If people have to create/do something different to make it work, when they upgrade, then we're doing something wrong.

And why do I need to add `http://localhost`, when Metabase is running on 3000? Seems like this is the problem, because when I add `http://localhost` together with `http://localhost:3000`, then it works.
I guess there's a bug in the React library (or the implementation), which doesn't correctly handle non-default ports.

**Comment by gusaiani** [comment]
> And why do I need to add http://localhost, when Metabase is running on 3000? Seems like this is the problem, because when I add http://localhost together with http://localhost:3000, 

@flamber, I don't really know why, and it seems buggy to me too.
I experimented a lot to get to this library, including doing everything vanilla. I vaguely recall having to do the same in the vanilla installation.

Anyway, I'll ask the library creator, he's been pretty responsive and nice.

**Comment by gusaiani** [comment]
> So the same steps as currently? https://www.metabase.com/docs/latest/administration-guide/10-single-sign-on.html#enabling-google-sign-in

@flamber, I've been comparing before/after installation steps, so that this doc is helpful on update and for new customers.
Working to get a PR on this doc page, and will show it to you asap.

Obviously I'll be very grateful if you have suggestions, then or now.

File: src/metabase/server/middleware/security.clj (hunk 1 above)
Comment (gusaiani) [reply src/metabase/server/middleware/security.clj:54]: @flamber will we worthwhile looking in detail. Perhaps in a follow-up PR?

**Comment by gusaiani** [comment]
@flamber, on this Google credential thing.

In https://developers.google.com/identity/gsi/web/guides/migration

In https://developers.google.com/identity/gsi/web/guides/migration#a_sign-in_migration_example

There is this very important phrase: *In most cases you can simply continue to use your existing web application Client ID credentials. *

It's kind of lost in there, but it's very important.

In this case I suggest we merge and just test `stats` without changing anything. It may just work.

**Comment by gusaiani** [comment]
@flamber Sure, to be clear I'm not saying we should hurry to merge.

What I mean is once we merge, we don't update the Google API secret.
And we check if things continue to work.

In parallel, regarding the ports, I've asked the lib author: https://github.com/MomenSherif/react-oauth/issues/23
He flambered the issue with tags and assigns, I'm hopeful he will answer soon.

**Comment by flamber** [comment]
Well, that was the link I needed: https://developers.google.com/identity/gsi/web/guides/get-google-api-clientid
Then let's try merging this and see what happens. It should work without Console changes on most setups - otherwise it's going to be problematic.
